prusti-specs/src/extern_spec_rewriter/common.rs

```diff
@@ -1,27 +1,138 @@
+///
+/// # Example
+/// Given an external specification such as
+/// ```ignore
+/// #[extern_spec]
+/// impl SomeTrait for SomeStruct {
+///     // specs
+///     fn the_trait_method(&self, arg: Self::AssocType) -> Bar;
+/// }
+/// ```
+///
+/// Generates a stub method and *sanitized* spec functions:
+/// ```ignore
+/// // spec functions with "self" rewritten to "_self: SomeStruct"
+/// fn the_trait_method(_self: SomeStruct, arg: <SomeStruct as SomeTrait::AssocType> -> Bar {
+///     <SomeStruct as SomeTrait>::the_trait_method(_self, arg);
+///     unimplemented!()
+/// }
+/// ```
+///
+pub(crate) fn generate_extern_spec_method_stub<T: HasSignature + HasAttributes + Spanned>(
+    method: &T,
+    self_type: &syn::TypePath,
+    self_type_trait: Option<&syn::TypePath>,
+    extern_spec_kind: ExternSpecKind,
+) -> syn::Result<(syn::ImplItemMethod, Vec<syn::ImplItemMethod>)> {
+    let method_sig = method.sig().clone();
+    let method_sig_span = method_sig.span();
+    let method_ident = &method_sig.ident;
+
+    // Determine path to externally specified method in UFCS
+    let method_path: syn::ExprPath = match self_type_trait {
+        Some(self_type_as_trait) => parse_quote_spanned! {method_sig_span=>
+            <#self_type as #self_type_as_trait> :: #method_ident
+        },
+        None => parse_quote_spanned! {method_sig_span=>
+            <#self_type> :: #method_ident
+        }
+    };
+
+    // Build the method stub
+    let method_attrs = method.attrs().clone();
+    let method_args = &method_sig.params_as_call_args();
+    let extern_spec_kind_string: String = extern_spec_kind.into();
+    let stub_method: syn::ImplItemMethod = parse_quote_spanned! {method.span()=>
+        #[trusted]
+        #[prusti::extern_spec = #extern_spec_kind_string]
+        #(#method_attrs)*
+        #[allow(unused, dead_code)]
+        #method_sig {
+            #method_path ( #method_args );
+            unimplemented!()
+        }
+    };
+
+    // Eagerly extract and process specifications
+    let mut stub_method = AnyFnItem::ImplMethod(stub_method);
+    let prusti_attributes = extract_prusti_attributes(&mut stub_method);
+    let (spec_items, generated_attributes) =
+        generate_spec_and_assertions(prusti_attributes, &stub_method)?;
+
+    // In the generated spec items and the stub method:
+    // - Rewrite associated types
+    // - Rewrite "self" to "_self"
+    let self_type_path = parse_quote_spanned! {self_type.span()=> #self_type };
+
+    let mut stub_method = stub_method.expect_impl_item();
+    stub_method.attrs.extend(generated_attributes);
+    stub_method.rewrite_self_type(&self_type_path, self_type_trait);
+    stub_method.rewrite_receiver(&self_type_path);
+
+    // Set span of generated method to externally specified method for better error reporting
+    syn::visit_mut::visit_impl_item_method_mut(&mut SpanOverrider::new(method_sig_span), &mut stub_method);
```

Comment:
Which spans will this override at this point? Is it still necessary?

Reply:
I re-added this line because some tests were failing.